lib/AST/ASTPrinter.cpp

```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  if (expr->isObjC()) {
+    Printer << "#keyPath(";
+  }
+  
+  if (!expr->isObjC()) {
+    if (auto rootType = expr->getRootType()) {
+      Printer << "\\";
+      printTypeLoc(rootType);
+    } else {
+      visit(expr->getParsedRoot());
+    }
+  }
+  
+  auto components = expr->getComponents();
+  if (!components.empty()) {
+    for (auto& component: components) {
+      auto kind = component.getKind();
+      switch (kind) {
+        case swift::KeyPathExpr::Component::Kind::Invalid: {
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::UnresolvedProperty: {
+          Printer << "." << component.getUnresolvedDeclName();
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::UnresolvedSubscript: {
+          auto args = component.getSubscriptArgs();
+          Printer << ".";
```

Comment:
KeyPath subscripts _can_ appear with a dot for the first component, but otherwise they need to be without the dot. I think it's probably best to print without the dot in all cases though.